src/bio/bio_dev_monitor.c

```diff
@@ -0,0 +1,442 @@
+	D_PRINT("\n");
+	D_PRINT("Unrecoverable Media Errors:  ");
+	dprint_uint128_dec(health_page->media_errors);
+	D_PRINT("\n");
+	D_PRINT("Lifetime Error Log Entries:  ");
+	dprint_uint128_dec(health_page->num_error_info_log_entries);
+	D_PRINT("\n");
+	D_PRINT("Warning Temperature Time:    %u minutes\n",
+		health_page->warning_temp_time);
+	D_PRINT("Critical Temperature Time:   %u minutes\n",
+		health_page->critical_temp_time);
+
+	/* Prep NVMe command to get controller data */
+	ctrlr_page_sz = sizeof(struct spdk_nvme_ctrlr_data);
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_IDENTIFY;
+	cmd.cdw10 = SPDK_NVME_IDENTIFY_CTRLR;
+
+	/*
+	 * Submit an NVMe Admin command to get controller data
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(health_desc, channel, &cmd,
+					   dev_health->bhm_ctrlr_buf,
+					   ctrlr_page_sz,
+					   get_spdk_identify_ctrlr_completion,
+					   dev_health);
+	if (rc)
+		D_ERROR("NVMe admin passthru (identify ctrlr), rc:%d\n", rc);
+}
+
+/* Get the SPDK device health state log and print all useful stats */
+void
+query_spdk_health_stats(struct bio_xs_context *ctxt, uint64_t now)
+{
+	struct bio_health_monitoring	*dev_health;
+	struct spdk_bdev		*bdev;
+	struct spdk_io_channel		*channel;
+	struct spdk_bdev_desc		*health_desc;
+	struct spdk_nvme_cmd		 cmd;
+	uint64_t			 health_stat_age;
+	int				 rc;
+	uint32_t			 numd, numdl, numdu;
+	uint32_t			 health_page_sz;
+
+	D_ASSERT(ctxt != NULL);
+	D_ASSERT(ctxt->bxc_blobstore != NULL);
+
+	dev_health = ctxt->bxc_blobstore->bb_dev_health;
+	health_stat_age = dev_health->bhm_stat_age;
+	/*
+	 * TODO Decide on an appropriate period to query device health
+	 * stats. Currently set at 1 min.
+	 */
+	if (health_stat_age + DAOS_SPDK_STATS_PERIOD >= now)
+		return;
+	dev_health->bhm_stat_age = now;
+
+
+	if (!dev_health->bhm_desc)
+		return;
+	health_desc = dev_health->bhm_desc;
+
+	bdev = spdk_bdev_desc_get_bdev(health_desc);
+	if (bdev == NULL) {
+		D_ERROR("Failed to get bdev from health desc:%p\n",
+			health_desc);
+		return;
+	}
+
+	if (!spdk_bdev_io_type_supported(bdev, SPDK_BDEV_IO_TYPE_NVME_ADMIN)) {
+		D_ERROR("Bdev NVMe admin passthru not supported!\n");
+		return;
+	}
+
+	channel = spdk_bdev_get_io_channel(health_desc);
+	if (channel == NULL) {
+		D_ERROR("Failed to get bdev I/O channel from health desc:%p\n",
+			health_desc);
+		return;
+	}
+
+	/* Prep NVMe command to get SPDK device health data */
+	health_page_sz = sizeof(struct spdk_nvme_health_information_page);
+	numd = health_page_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd.cdw10 = numdl << 16;
+	cmd.cdw10 |= SPDK_NVME_LOG_HEALTH_INFORMATION;
+	cmd.cdw11 = numdu;
+
+	/*
+	 * Submit an NVMe Admin command to get device health log page
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(health_desc, channel, &cmd,
```

Comment:
We issue health query every minute, and the admin completion is polled every second, it's unlikely that two query cmd are issued in parallel, but we'd track the in-flight query and avoid issue another query before the previous one done.

Reply:
would it be enough to have it simply check if there is a current device query inflight, and if there is just return (since it is probably not needed to have 2 back-to-back queries if the case arises where 2 query commands are issued in parallel)? or would you think a bust wait or something is needed instead to wait until the query is complete to continue?

Reply:
Yes, simply check to avoid 2 back-to-back queries will be fine, we can't afford wait here.

Reply:
done